controllers/awsmachine_controller.go

```diff
@@ -410,8 +411,13 @@ func (r *AWSMachineReconciler) reconcileNormal(_ context.Context, machineScope *
 		return ctrl.Result{}, nil
 	}
 
+	failureDomain := machineScope.AWSMachine.Spec.FailureDomain
+	if failureDomain == nil {
+		failureDomain = machineScope.Machine.Spec.FailureDomain
+	}
```

Comment:
Now this logic and the format `fmt.Sprintf("aws:///%s/%s", aws.StringValue(failureDomain)`  are invariant. If we move them inside the `SetProviderID()` func and we pass only the instance.ID we can unit test the multiple scenarios for failureDomain, e.g both empty, one empty, both filled. wdyt

Reply:
Good idea, that's definitely easier to test